aten/src/ATen/TensorIterator.cpp

```diff
@@ -653,20 +670,30 @@ void TensorIteratorBase::for_each(loop2d_t loop, int64_t grain_size) {
 }
 
 StrideVector TensorIteratorBase::get_strides() const {
-  StrideVector strides;
-  for (int dim = 0; dim < ndim(); dim++) {
-    for (int arg = 0; arg < ntensors(); arg++) {
-      strides.push_back(operands_[arg].stride_bytes[dim]);
-    }
-  }
+  StrideVector strides(ndim() * ntensors());
+  at::get_strides(strides.data(), operands_, ndim());
   return strides;
 }
 
 void TensorIteratorBase::serial_for_each(loop2d_t loop, Range range) const {
   if (range.size() == 0) {
     return;
   }
-  at::internal::serial_for_each(shape_, get_strides(), get_base_ptrs(), loop, range);
+
+  const auto ntensors = this->ntensors();
+  const auto ndim = this->ndim();
+
+  c10::SmallBuffer<char*, 4> ptrs(ntensors);
+  c10::SmallBuffer<int64_t, 8> strides(ntensors * std::max(2, ndim));
+
+  at::get_base_ptrs(ptrs.data(), operands_);
+  at::get_strides(strides.data(), operands_, ndim);
+  if (ndim < 2) {
+    std::fill(strides.begin() + ntensors * ndim, strides.end(), 0);
+  }
```

Comment:
Zero-padding here costs around 800 instructions, so still not free but still cheaper than the way it was done previously.